Add option to disable adding trailing slash

MateuszWkDev opened the pull request:
Middlewares affected:
- DefaultFilesMiddleware
- DirectoryBrowserMiddleware

Added AppendTrailingSlash property with default value= true in SharedOptions for files middleware. Now users are able to disable redirect to path with trailing slash. Also added some test cases for this option changed to false.

Closes #2449

MateuszWk committed:
Add option to disable adding trailing slash

Middlewares affected:
- DefaultFilesMiddleware
- DirectoryBrowserMiddleware

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/aspnet/AspNetCore?pullRequest=12669) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/aspnet/AspNetCore?pullRequest=12669) before we can accept your contribution.<br/><br/>:x: MateuszWkDev [sign now](https://cla.dotnetfoundation.org/aspnet/AspNetCore?pullRequest=12669)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/aspnet/AspNetCore?pullRequest=12669) it.</sub>

MateuszWk committed:
Reference assemblies

MateuszWk committed:
Pass param to FoundDirectoryWithDefaultFile_PathModified

MateuszWk committed:
Add proper test cases for FoundDirectoryWithDefaultFile_PathModified_All

Tratcher commented on src/Middleware/StaticFiles/src/Helpers.cs:
```diff
@@ -28,6 +31,24 @@ internal static bool PathEndsInSlash(PathString path)
+            return path.Value;
+        }
+
+        internal static Task RedirectToPathWithSlash(HttpContext context)
```
Task -> void.

MateuszWkDev replied:
changed in Resolve suggestions

Tratcher commented on src/Middleware/StaticFiles/test/UnitTests/DefaultFilesMiddlewareTests.cs:
```diff
@@ -213,24 +227,34 @@ public async Task PostDirectory_PassesThrough_All(string baseUrl, string baseDir
+        [Fact]
+        public void Options_AppendTrailingSlashByDefault()
+        {
+            Assert.True(new DefaultFilesOptions().AppendTrailingSlash);
```
And directory browser.

MateuszWkDev replied:
I check DirectoryBrowserOptions in DirectoryBrowserMiddlewareTests. Test logic is the same

Tratcher commented on src/Middleware/StaticFiles/test/UnitTests/DefaultFilesMiddlewareTests.cs:
```diff
@@ -213,24 +227,34 @@ public async Task PostDirectory_PassesThrough_All(string baseUrl, string baseDir
         }
 
-        private async Task PostDirectory_PassesThrough(string baseUrl, string baseDir, string requestUrl)
+        private async Task PostDirectory_PassesThrough(string baseUrl, string baseDir, string requestUrl, bool appendTrailingSlash = true)
```
All of the test scenarios you've added have the same result with and without the new option. There needs to be a test that confirms a redirect is not generated with the new option changed. See NearMatch_RedirectAddSlash.

MateuszWkDev replied:
I've added new inline data without trailing slash at the end, if RedirectToAppendTrailingSlash wouldn't work those test would fail beacause of redirect, each test checks response.StatusCode.

Tratcher commented on src/Middleware/StaticFiles/src/Infrastructure/SharedOptions.cs:
```diff
@@ -42,5 +42,10 @@ public PathString RequestPath
+        /// <summary>
+        /// Indicates whether to add trailing slash at the end of path
+        /// </summary>
+        public bool AppendTrailingSlash { get; set; } = true;
```
You're trying to suppress the redirect, not the trailing slash, right? You still add the trailing slash internally.
RedirectToAppendTrailingSlash?

MateuszWkDev replied:
Renamed in next commit

Tratcher commented on src/Middleware/StaticFiles/src/Infrastructure/SharedOptions.cs:
```diff
@@ -42,5 +42,10 @@ public PathString RequestPath
         public IFileProvider FileProvider { get; set; }
+
+        /// <summary>
+        /// Indicates whether to add trailing slash at the end of path
```
```suggestion
        /// Indicates whether to redirect to add a trailing slash at the end of path. Relative resource links may require this.
```

MateuszWkDev replied:
changed

MateuszWkDev replied:
Renamed according to your suggestion in next commit

MateuszWk committed:
Resolve suggestions

- change property name to RedirectToAppendTrailingSlash
- update summary for RedirectToAppendTrailingSlash
- return void in RedirectToPathWithSlash
- add more test cases

Tratcher commented:
Thanks
